Consolidate event history docs and modernize Continue as New encyclopedia (#3573)
temporalio/documentation: docs/encyclopedia/workflow/workflow-execution/continue-as-new.mdx

  - Workflows
---

This page discusses [Continue-As-New](#continue-as-new) and how to tell [when](#when) you would want to use it.

## What is Continue-As-New? {#continue-as-new}

Continue-As-New allows you to checkpoint your Workflow's state and start a fresh Workflow.

> There are two main problems this practice can prevent:

Reviewer: ```suggestion
There are main reasons you might want to do this:
```

Reviewer 2: ```suggestion
There are two main reasons you might want to start a new Workflow:
```
Adding a little specificity; ambiguous use of "this" jumps out at me.

- A Workflow Execution with a long, or large [Event History](/workflow-execution/event#event-history), such as one calling many activities, may bog down and have performance issues.
  It could even generate more events than allowed by the [Event History limits](/workflow-execution/event#event-history-limits).
- A Workflow Execution can hit [Workflow Versioning](/workflow-definition#workflow-versioning) problems if it started running on an older version of your code.

Pass your latest relevant state into Continue-As-New.
This checkpoints it and passes it to a new Execution in the [Execution Chain](/workflow-execution#workflow-execution-chain).
This state is passed in through your Workflow's parameters.

> The parameters are typically optional and left unset by the original caller of the Workflow.

Reviewer: I'd probably get rid of "This checkpoints it" because as a reader that would make me think that Temporal had a thing called "checkpointing" that I didn't know about.

I'm also tempted to add a sentence saying something like

> The end result is that you've created a new workflow with a fresh history, that picks up where your last one left off.

The new Workflow Execution has the same Workflow Id, but a different Run Id, and starts its own Event History.

> You can repeat Continue-As-New as often as needed, which means that your Workflow can run in perpetuity.

Reviewer: Channeling the docs team I wonder if "can run forever" will be more widely understood without losing much in accuracy.

Workflows that do this are often called Entity Workflows because they represent durable objects, not just processes.

- [How to Continue-As-New using the Go SDK](/develop/go/continue-as-new#how)
- [How to Continue-As-New using the Java SDK](/develop/java/continue-as-new#how)
- [How to Continue-As-New using the PHP SDK](/develop/php/continue-as-new#how)
- [How to Continue-As-New using the Python SDK](/develop/python/continue-as-new#how)
- [How to Continue-As-New using the TypeScript SDK](/develop/typescript/continue-as-new#how)
- [How to Continue-As-New using the .NET SDK](/develop/dotnet/continue-as-new#how)

## When is it right to Continue-As-New? {#when}

To prevent scalability problems, periodically check Continue As New Suggested in your Workflow.\
Temporal will tell your Workflow when it's approaching performance or scalability problems.

> To help prevent versioning problems for long-running Workflows, you may also want to Continue-as-New after a certain amount of time has elapsed. This will make sure every Workflow began on a recent version of your source code.

Reviewer: Not a versioning expert but this sounds a bit vague to me. After how much time? How do I know that a versioning error won't occur if the CAN is not issued at a precise point w.r.t. workflow code deploys?

Author: Heh, we're working on a better answer here--Trampolining.  So yeah, we don't have a satisfying answer ATM.  Some users have used signals.

Author: This really is the most practical advice for now, since very few users have chosen to use signals.  My goal in this doc is to limit it to a few versions rather than one version.

Author: Wordsmithed.

> - [Determine when to Continue-As-New using the Go SDK](/develop/go/continue-as-new#when)

Author: Note: Other than python, these #when links don't exist yet, so they will link to the top of those pages until we finish those docs.

  - Workflows
---

> This page discusses [Continue-As-New](#continue-as-new) and how to tell [when](#when) you would want to use it.

Reviewer 2: ```suggestion
This page discusses [Continue-As-New](#continue-as-new) and how to decide [when to use it](#when).
```

First, pass your latest relevant state into Continue-As-New.
This hands it to a new Execution in the [Execution Chain](/workflow-execution#workflow-execution-chain).
This state is passed in as arguments to your Workflow.
The parameters are typically optional and left unset by the original caller of the Workflow.

The new Workflow Execution has the same Workflow Id, but a different Run Id, and starts its own Event History.

You can repeat Continue-As-New as often as needed, which means that your Workflow can run forever.
Workflows that do this are often called Entity Workflows because they represent durable objects, not just processes.

- [How to Continue-As-New using the Go SDK](/develop/go/continue-as-new#how)
- [How to Continue-As-New using the Java SDK](/develop/java/continue-as-new#how)
- [How to Continue-As-New using the PHP SDK](/develop/php/continue-as-new#how)
- [How to Continue-As-New using the Python SDK](/develop/python/continue-as-new#how)
- [How to Continue-As-New using the TypeScript SDK](/develop/typescript/continue-as-new#how)
- [How to Continue-As-New using the .NET SDK](/develop/dotnet/continue-as-new#how)

> ## When is it right to Continue-As-New? {#when}

Reviewer 2: How about "When should you use Continue-As-New?" Seems more to-the-point, but I could be missing something.

Author: I wanted it to be clear that the question was _at what point_ in the execution to continue as new rather than  _whether_ to use it.  And "when should you" is fairly ambiguous.

> To prevent scalability problems, periodically check Continue As New Suggested in your Workflow.\

Reviewer 2: ```suggestion
To prevent scalability problems, periodically check Continue As New Suggested in your Workflow.
```
Where would the user see the "Continue As New Suggested" message? I'm trying to describe this something like "periodically check for the Continue As New Suggested message in the CLI" or however it works.

> Temporal will tell your Workflow when it's approaching performance or scalability problems.

Reviewer 2: I suggest swapping lines 48 and 49. Tell the reader about this cool thing Temporal does, then tell them what to look for specifically.

> To prevent long-running Workflows from running on stale versions of code, you may also want to Continue-as-New periodically. This makes sure you're running only a couple of versions, which avoids some backwards compatibility problems.

Reviewer 2: Can we define "periodically," or at least give some kind of guidelines? Weekly, monthly, every X number of executions?

Author: There's no concrete time frame because it depends on how frequently they deploy, which varies widely.
